libethereum/Block.cpp

```diff
@@ -136,47 +136,59 @@ void Block::noteChain(BlockChain const& _bc)
     }
 }
 
-PopulationStatistics Block::populateFromChain(BlockChain const& _bc, h256 const& _h, ImportRequirements::value _ir)
+void Block::populateFromChain(BlockChain const& _bc, h256 const& _h)
 {
-    noteChain(_bc);
-
-    PopulationStatistics ret { 0.0, 0.0 };
-
     if (!_bc.isKnown(_h))
     {
-        // Might be worth throwing here.
-        cwarn << "Invalid block given for state population: " << _h;
+        cerr << "Invalid block given for state population: " << _h;
         BOOST_THROW_EXCEPTION(BlockNotFound() << errinfo_target(_h));
     }
 
-    auto b = _bc.block(_h);
-    BlockHeader bi(b);		// No need to check - it's already in the DB.
-    if (bi.number())
+    auto const& blockBytes = _bc.block(_h);
+    m_currentBytes = blockBytes;
+
+    // Set block headers
+    auto const& blockHeader = BlockHeader{blockBytes};
+    m_currentBlock = blockHeader;
+
+    if (blockHeader.number())
+        m_previousBlock = _bc.info(blockHeader.parentHash());
+    else
+        m_previousBlock = m_currentBlock;
+
+    // Set state root and precommit state
+    //
+    // First check for database corruption by looking up the state root in the state database. Note
+    // that we don't technically need to do this since if the state DB is corrupt setting a new
+    // state root will throw anyway, but checking here enables us to log a user-friendly error
+    // message.
+    if (m_state.db().lookup(blockHeader.stateRoot()).empty())
     {
-        // Non-genesis:
-
-        // 1. Start at parent's end state (state root).
-        BlockHeader bip(_bc.block(bi.parentHash()));
-        sync(_bc, bi.parentHash(), bip);
-
-        // 2. Enact the block's transactions onto this state.
-        m_author = bi.author();
-        Timer t;
-        auto vb = _bc.verifyBlock(&b, function<void(Exception&)>(), _ir | ImportRequirements::TransactionBasic);
-        ret.verify = t.elapsed();
-        t.restart();
-        enact(vb, _bc);
-        ret.enact = t.elapsed();
+        cerr << "Unable to populate block " << blockHeader.hash() << " - state root "
+             << blockHeader.stateRoot() << " not found in database.";
+        cerr << "Database corrupt: contains block without state root: " << blockHeader;
+        cerr << "Try rescuing the database by running: eth --rescue";
+        BOOST_THROW_EXCEPTION(InvalidStateRoot() << errinfo_target(blockHeader.stateRoot()));
     }
-    else
+
+    m_state.setRoot(blockHeader.stateRoot());
+    m_precommit = m_state;
+
+    RLP blockRLP{blockBytes};
+    auto const& txListRLP = blockRLP[1];
+    for (auto const& txRLP : txListRLP)
     {
-        // Genesis required:
-        // We know there are no transactions, so just populate directly.
-        m_state = State(m_state.accountStartNonce(), m_state.db(), BaseState::Empty);	// TODO: try with PreExisting.
-        sync(_bc, _h, bi);
+        m_transactions.push_back(Transaction{txRLP.data(), CheckTransaction::None});
```

Comment:
Did it fill `m_transactions`, `m_transactionsSet`, `m_currentTxs`, `m_currentUncles` before?

I suspect these are also useful only for mining, and RPC methods get them separately from `BlockChain`

Reply:
Good question 😀 
* `m_transactions `& `m_transactionsSet`: Yes
* `m_currentTxs `& `m_currentUncles`: No

So you raise a good question - can we assume that this function will only be called via RPC? I populated everything in the block because `populateFromChain `is a public function so it can be called by anyone...unless these fields are only usable in very specific situations (e.g. mining), perhaps we should fork the `populateFromChain `into two functions - 1 which populates the fields which are useful in most cases and another which populates the rest?

Reply:
Yeah I think ideally these different cases should be handled somehow by different classes.

But let's not complicate it here, I think I'm fine with leaving like this.
But `m_currentTxs` really looks to be specific to only `commitToSeal` and `sealBlock` methods, and I think it would be fine to not copy  it here. `m_currentUncles` probably too.